File: 백준/Silver/1149. RGB거리/RGB거리.py
import sys
input = sys.stdin.readline

n = int(input())
house = []
cost = [[0, 0, 0] for _ in range(n)]

for _ in range(n):
    house.append(list(map(int, input().split())))

# 첫 번째 집은 첫 번째 비용으로 초기화
cost[0][0] = house[0][0]
cost[0][1] = house[0][1]
cost[0][2] = house[0][2]

# 이전 집과 겹치지 않는 색상 중에서 이전 집까지 더한 비용과 현재 색상 비용을 더해서 최소값을 할당
for i in range(1, n):
    cost[i][0] = min(house[i][0] + cost[i-1][2], house[i][0] + cost[i-1][1])
    cost[i][1] = min(house[i][1] + cost[i-1][2], house[i][1] + cost[i-1][0])
    cost[i][2] = min(house[i][2] + cost[i-1][0], house[i][2] + cost[i-1][1])

print(min(cost[n-1]))

File: 백준/Silver/1149. RGB거리/RGB거리.java
import java.util.*;
import java.io.*;

public class Main {
	static BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	static StringBuilder output = new StringBuilder();
	static StringTokenizer tokens;

	public static void main(String[] args) throws IOException {
		int N = Integer.parseInt(input.readLine());
		int[][] rgb = new int[N][3];
		int[][] dp = new int[N][3];
		
		for(int r = 0; r < N; r++) {
			tokens = new StringTokenizer(input.readLine());
			for(int c = 0; c < 3; c++) {
				rgb[r][c] = Integer.parseInt(tokens.nextToken());
			}
		}

		
		for(int c = 0; c < 3; c++) {
			dp[0][c] = rgb[0][c];
		}
		
		for(int r = 1; r < N; r++) {			
			dp[r][0] = rgb[r][0] + Math.min(dp[r-1][1], dp[r-1][2]);
			dp[r][1] = rgb[r][1] + Math.min(dp[r-1][0], dp[r-1][2]);
			dp[r][2] = rgb[r][2] + Math.min(dp[r-1][1], dp[r-1][0]);
		}
		
		System.out.println(Math.min(Math.min(dp[N-1][0], dp[N-1][1]), dp[N-1][2]));
	}

}